### DISCUSSION TITLE
Error with mesh created with gmsh: "ERROR: Expected boundary element of dimension 2 but.."

### DISCUSSION THREAD

**Discussion by richmondodufisan** [idaholab/moose#25160] [Q&A Meshing] [answered]
I am trying to run a problem with the heat conduction module, and I got an error from MOOSE:

```
*** ERROR ***
ERROR: Expected boundary element of dimension 2 but got 1
Stack frames: 17
0: libMesh::print_trace(std::ostream&)
1: libMesh::MacroFunctions::report_error(char const*, int, char const*, char const*, std::ostream&)
2: libMesh::AbaqusIO::assign_sideset_ids()
3: libMesh::AbaqusIO::read(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
4: libMesh::NameBasedIO::read(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
5: libMesh::UnstructuredMesh::read(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, void*, bool, bool)
6: FileMesh::buildMesh()
7: MooseMesh::init()
8: SetupMeshAction::act()
9: Action::timedAct()
10: ActionWarehouse::executeActionsWithAction(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
11: ActionWarehouse::executeAllActions()
12: MooseApp::runInputFile()
13: MooseApp::run()
14: main
15: __libc_start_main
16: ./farm-opt(+0x33ef) [0x55cd03a3d3ef]
[0] ../src/mesh/abaqus_io.C, line 1168, compiled Apr 20 2023 at 15:32:41
```

The error message unfortunately doesn't reference anything in my input file, so I don't know exactly where to look. To describe the mesh, I have two blocks (volumes) stacked on top of each other, with an interface between them. I defined each block in gmsh as a volume, labeled the outer surfaces of the mesh, and also labeled the interface between them. I also have a sphere on top which is just for refining an area of interest. Here is the gmsh .geo script that produced my mesh:

```
// Gmsh project created on Wed Aug 09 07:54:14 2023
xcen = 0;
ycen = 0;
radius = 8;
dop_thick = 0.09;
pump_refine = 0.3;
reg_element = 2;

SetFactory("OpenCASCADE");
//+
Point(1) = {40, 20, 0, reg_element};
//+
Point(2) = {40, -20, 0, reg_element};
//+
Point(3) = {-40, -20, 0, reg_element};
//+
Point(4) = {-40, 20, 0, reg_element};
//+
Point(5) = {-40, 20, -40, reg_element};
//+
Point(6) = {-40, -20, -40, reg_element};
//+
Point(7) = {40, -20, -40, reg_element};
//+
Point(8) = {40, 20, -40, reg_element};
//+
Line(1) = {3, 2};
//+
Line(2) = {3, 6};
//+
Line(3) = {6, 7};
//+
Line(4) = {7, 2};
//+
Line(5) = {6, 5};
//+
Line(6) = {5, 8};
//+
Line(7) = {8, 7};
//+
Line(8) = {8, 1};
//+
Line(9) = {1, 2};
//+
Line(10) = {2, 3};
//+
Line(11) = {3, 4};
//+
Line(12) = {4, 5};
//+
Line(13) = {4, 1};
//+
Curve Loop(1) = {11, 12, -5, -2};
//+
Plane Surface(1) = {1};
//+
Curve Loop(2) = {3, 4, -1, 2};
//+
Plane Surface(2) = {2};
//+
Curve Loop(3) = {7, 4, -9, -8};
//+
Plane Surface(3) = {3};
//+
Curve Loop(4) = {6, 7, -3, 5};
//+
Plane Surface(4) = {4};
//+
Curve Loop(5) = {6, 8, -13, 12};
//+
Plane Surface(5) = {5};
//+
Curve Loop(6) = {1, -9, -13, -11};
//+
Plane Surface(6) = {6};
//+
Point(9) = {40, 20, dop_thick, reg_element};
//+
Point(10) = {40, -20, dop_thick, reg_element};
//+
Point(11) = {-40, -20, dop_thick, reg_element};
//+
Point(12) = {-40, 20, dop_thick, reg_element};
//+
Line(14) = {9, 10};
//+
Line(15) = {10, 11};
//+
Line(16) = {11, 12};
//+
Line(17) = {12, 9};
//+
Line(18) = {11, 3};
//+
Line(19) = {12, 4};
//+
Line(20) = {10, 2};
//+
Line(21) = {9, 1};
//+
Curve Loop(7) = {11, -19, -16, 18};
//+
Plane Surface(7) = {7};
//+
Curve Loop(8) = {15, 18, 1, -20};
//+
Plane Surface(8) = {8};
//+
Curve Loop(9) = {9, -20, -14, 21};
//+
Plane Surface(9) = {9};
//+
Curve Loop(10) = {13, -21, -17, 19};
//+
Plane Surface(10) = {10};
//+
Curve Loop(11) = {15, 16, 17, 14};
//+
Plane Surface(11) = {11};
//+
Surface Loop(1) = {2, 4, 5, 3, 1, 8, 11, 7, 10, 9};
//+
Surface Loop(2) = {6, 2, 4, 5, 3, 1};
//+
Volume(1) = {2};
//+
Surface Loop(3) = {6, 11, 8, 7, 10, 9};
//+
Volume(2) = {3};
//+
Point(13) = {xcen, ycen, dop_thick, pump_refine};
//+
Point(14) = {xcen, ycen+radius, dop_thick, pump_refine};
//+
Point(15) = {xcen, ycen-radius, dop_thick, pump_refine};
//+
Point(16) = {xcen+radius, ycen, dop_thick, pump_refine};
//+
Point(17) = {xcen-radius, ycen, dop_thick, pump_refine};
//+
Point(18) = {xcen, ycen, dop_thick-radius, pump_refine};
//+
Circle(22) = {17, 13, 15};
//+
Circle(23) = {15, 13, 16};
//+
Circle(24) = {16, 13, 14};
//+
Circle(25) = {17, 13, 14};
//+
Circle(26) = {17, 13, 18};
//+
Circle(27) = {15, 13, 18};
//+
Circle(28) = {16, 13, 18};
//+
Circle(29) = {14, 13, 18};
//+
Curve Loop(12) = {22, 23, 24, -25};
//+
Plane Surface(12) = {12};
//+
Curve Loop(13) = {26, -27, -22};
//+
Surface(13) = {13};
//+
Curve Loop(15) = {23, 28, -27};
//+
Surface(14) = {15};
//+
Curve Loop(17) = {24, 29, -28};
//+
Surface(15) = {17};
//+
Curve Loop(19) = {25, 29, -26};
//+
Surface(16) = {19};
//+
Physical Surface("boundary_conductance", 30) = {6};
//+
Physical Surface("outer_surface", 31) = {1, 4, 3, 2, 5, 7, 8, 9, 10};
//+
Physical Surface("top_heat_source", 32) = {11};
//+
Physical Volume("sample_material", 33) = {1};
//+
Physical Volume("doping_material", 34) = {2};
```

Also, when exporting the mesh, I exported it as an inp file and selected both options (for those familiar with gmsh: "save all elements" and "save all groups of nodes"). If I uncheck either, MOOSE could not find the blocks mentioned in the code. I also have "construct_side_list_from_node_list=true", it doesn't affect the error. What could be the issue?

**Comment by lindsayad** [comment]
Files saved with the `.inp` extension are processed as Abaqus files. If you want the gmsh reader you will need to use the `.msh` extension

**Reply by richmondodufisan** [reply]
So, gmsh can apparently export files in the Abaqus format. However, I tried exporting in the gmsh format and still got an error:

```
*** ERROR ***
map_find() error: key "26" not found in file ../src/mesh/gmsh_io.C on line 508
Stack frames: 18
0: libMesh::print_trace(std::ostream&)
1: libMesh::MacroFunctions::report_error(char const*, int, char const*, char const*, std::ostream&)
2: std::map<unsigned int, libMesh::GmshIO::ElementDefinition, std::less<unsigned int>, std::allocator<std::pair<unsigned int const, libMesh::GmshIO::ElementDefinition> > >::mapped_type& libMesh::Utility::map_find<std::map<unsigned int, libMesh::GmshIO::ElementDefinition, std::less<unsigned int>, std::allocator<std::pair<unsigned int const, libMesh::GmshIO::ElementDefinition> > >, unsigned int, (unsigned int*)0>(std::map<unsigned int, libMesh::GmshIO::ElementDefinition, std::less<unsigned int>, std::allocator<std::pair<unsigned int const, libMesh::GmshIO::ElementDefinition> > >&, unsigned int const&, char const*, int)
3: libMesh::GmshIO::read_mesh(std::istream&)
4: libMesh::GmshIO::read(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
5: libMesh::NameBasedIO::read(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
6: libMesh::UnstructuredMesh::read(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, void*, bool, bool)
7: FileMesh::buildMesh()
8: MooseMesh::init()
9: SetupMeshAction::act()
10: Action::timedAct()
11: ActionWarehouse::executeActionsWithAction(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
12: ActionWarehouse::executeAllActions()
13: MooseApp::runInputFile()
14: MooseApp::run()
15: main
16: __libc_start_main
17: ./farm-opt(+0x33ef) [0x5599600983ef]
[0] ./include/libmesh/utility.h, line 134, compiled Apr 20 2023 at 15:32:43
```

What could this be referring to?

**Reply by 1runer** [reply]
Did you try to load the *.msh file using your app with --mesh-only options?
Is an error message also displayed? Or just when you try to run the entire simulation?

I don't have a suitable input file to use your mesh, but simply loading and creating an exodus works for me and can be viewed in paraview.

**Reply by richmondodufisan** [reply]
I used an updated version of gmsh (4.11, I was using 4.9) to make the mesh in .msh format and it seems like its working (no errors like before). But-

I just tried the --mesh-only flag and it did output a .e file which, when I viewed in paraview, was just a surface. When I selected "volume" it was empty, and I didn't see the spherical refinement I did (were you able to see it?)

Also, running the program again with --mesh-only gives the same output on the console but doesn't create a .e file anymore (it only created one the first time). 

Now I'm REALLY confused. I don't know if its my computer at this point.

**Reply by GiudGiud** [reply]
`--mesh-only` creates a `<input_name>_in.e` file

**Reply by richmondodufisan** [reply]
@1runer in the .e file output with the --mesh-only flag, do you get a solid volume mesh or just a surface? are you able to see the semisphere where I increased the mesh refinement?

**Reply by 1runer** [reply]
It looks to me that your mesh is not properly defined in the .geo-file.

Opening/Meshing with Gmsh shows to me two blocks: A big cube and attached a super thin cube. 

The sphere is in the geometry file but not in the meshed result.
Did you embed the internal surfaces into the 3D volume? This is needed to have a conforming mesh following the internal curves of the sphere.
Did you define a Physical Volume for the sphere? 

If both answers are no, you have simply created two unconnected meshes in Gmsh. And since the sphere must be defined as a Physical Volume, it is simply deleted in the output to .msh.

![image](https://github.com/idaholab/moose/assets/26140953/c0bd1c3a-36cb-430a-ab1f-735b0d4440ee)

**Reply by richmondodufisan** [reply]
@1runer 
That's right, I have one large cube and one super thin cube on top of it as you were able to see. I'm surprised that the semisphere didn't show up for you though. I just reloaded the .geo file with the script above and this is what I have:

![gmsh_mesh2](https://github.com/idaholab/moose/assets/52802063/f6f4f9e0-a164-4f36-b291-17d0b3dab91c)

I defined internal surfaces for the sphere but did not define a volume, as it is embedded within two existing volumes and isn't really a volume itself. Its a "dummy" shape for refinement of that specific area. This is the mesh I get after meshing:

![gmsh_mesh1](https://github.com/idaholab/moose/assets/52802063/ca4078bc-7b6d-4b68-b688-129880f34405)

But when I export this same thing to MOOSE, the output mesh I get is just a hollow surface cube when I view in paraview (not even a volume at all):

![paraview1](https://github.com/idaholab/moose/assets/52802063/5148402a-0caa-47eb-a1c0-61bacd6a9356)

 Do you have any idea what could be the problem here?

**Reply by 1runer** [reply, marked as answer]
If I click the mesh 3D button, I get a 3D mesh and a lot of warnings because of the bad aspect ratio of the small cube. But it meshes 3D.

You have to explicitly embed the lines/surfaces in the volume to have a conforming mesh. There is a command for doing it. 
Otherwise you create two unconnected meshes.
Gmsh just outputs parts of any mesh which are assigned to a physical volume.

Please see the [Gmsh tutorial](https://gmsh.info/doc/texinfo/gmsh.html#t15) how to do so.

**Reply by richmondodufisan** [reply]
@1runer after battling with this yesterday I'm convinced the issue in MOOSE was that the original mesh was unconnected. Thanks! However, using either BooleanFragments or Coherence in gmsh to make the mesh conform made the physical groups I initially defined disappear, and I unfortunately have not found a way to redefine those groups. I'll mark this as answered since the actual problem was figured out.
